apps/workshop/projects/CroSwap/modules/CroSwap.ts

```diff
@@ -0,0 +1,152 @@
+import { CroSwapFarms, CroSwapTokens, CroSwapToken, Reserves, UserInfo } from '../helpers/types';
+import { subgraph } from '../helpers/gql';
+import { FetchPoolsContext, TokenExtra, UserPosition } from '../../../../sandbox/src/types/module';
+import { Pool } from '@defiyield/sandbox';
+import { findToken } from '../../../../../packages/utils/array';
+import {
+  CROSWAP_FARMS,
+  GQL_GET_TOKENS,
+  CROSWAP_FARMS_URL,
+  CROSWAP_SUBGRAPH_URL,
+} from '../helpers/const';
+import farmsAbi from '../abis/farms.abi.json';
+import pairAbi from '../abis/pair.abi.json';
+import { BigNumber, FixedNumber } from 'ethers/lib/ethers';
+import { formatUnits } from 'ethers/lib/utils';
+import { fetchPoolsFromUrl, preloadTokensFromUrl } from '../helpers/util';
+
+export const CroSwap: ModuleDefinitionInterface = {
+  name: 'CroSwap',
+  chain: 'cronos',
+  type: 'staking',
+
+  /**
+   * Fetches the addresses of all involved tokens (supplied, rewards, borrowed, etc)
+   *
+   * @param context
+   * @returns Address[]
+   */
+  async preloadTokens(ctx) {
+    return await preloadTokensFromUrl(ctx, CROSWAP_SUBGRAPH_URL);
+  },
+
+  /**
+   * Returns full pool list
+   *
+   * @param context
+   * @returns Pool[]
+   */
+  async fetchPools(ctx: FetchPoolsContext) {
+    return await fetchPoolsFromUrl(ctx, ctx.tokens, CROSWAP_FARMS_URL);
+  },
+
+  /**
+   * Returns user positions for all pools
+   *
+   * @param ctx Context
+   * @returns UserPosition[]
+   */
+  async fetchUserPositions(ctx) {
+    const { pools, user, ethcall, ethcallProvider, ethers } = ctx;
+    const farms = new ethcall.Contract(CROSWAP_FARMS, farmsAbi);
+
+    const positions = pools.flatMap(async (pool) => {
+      const token0 = pool.supplied?.[0].token;
+      if (!token0) return void 0;
+
+      const token1 = pool.supplied?.[1].token;
+      if (!token1) return void 0;
+
+      const lpTokenContract = new ethcall.Contract(pool.id, pairAbi);
+
+      const [poolId, balance, reserves, totalSupply] = (await ethcallProvider.all([
+        farms.getPoolIdforLP(pool.id),
+        lpTokenContract.balanceOf(user),
+        lpTokenContract.getReserves(),
+        lpTokenContract.totalSupply(),
+      ])) as [BigNumber, BigNumber, Reserves, BigNumber];
+
+      const [userInfo, pending] = (await ethcallProvider.all([
+        farms.getUserInfo(poolId, user),
+        farms.pendingRewards(poolId, user),
+      ])) as [UserInfo, BigNumber];
+
+      // total balance of lp tokens in wallet & staked
+      const totalBalance = balance.add(userInfo.amount);
+
+      // dont continue if no balance
+      if (totalBalance.eq(0)) return void 0;
+
+      const reserve0 = FixedNumber.fromValue(reserves._reserve0, 18);
+      const reserve1 = FixedNumber.fromValue(reserves._reserve1, 18);
+
+      // get ratio of total supply
+      const ratioOfTotalSupply = FixedNumber.fromValue(totalBalance).divUnsafe(
+        FixedNumber.from(totalSupply),
+      );
```

Comment:
same as the other comment, if you could update these to use the injected ethers instead of importing directly, thanks!

```suggestion
      const reserve0 = ethers.FixedNumber.fromValue(reserves._reserve0, 18);
      const reserve1 = ethers.FixedNumber.fromValue(reserves._reserve1, 18);

      // get ratio of total supply
      const ratioOfTotalSupply = ethers.FixedNumber.fromValue(totalBalance).divUnsafe(
        ethers.FixedNumber.from(totalSupply),
      );
```